### PR TITLE
loads only valid previous descriptors

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- loads only valid previous descriptors

## Why are we making these changes?

Fallback support should look back in history to a valid service description and the lookup should not be cut short by an intermediate invalid description.

### CODE REVIEW

File: waiter/src/waiter/descriptor.clj [hunk 1]
```diff
@@ -232,30 +232,36 @@
     (-> (headers/split-headers (:headers request))
         (sd/merge-service-description-sources kv-store waiter-hostnames service-description-defaults token-defaults)
         (sd/merge-service-description-and-id kv-store service-id-prefix current-request-user metric-group-mappings
-                                             service-description-builder assoc-run-as-user-approved?)
+                                             service-description-builder assoc-run-as-user-approved? true)
         (sd/merge-suspended kv-store))))
 
 (defn descriptor->previous-descriptor
```
Comment (sradack) [inline waiter/src/waiter/descriptor.clj]: `descriptor->previous-descriptor` doesn't match what this function is doing anymore.   The goal of this change is to change our fallback descriptors.  We have a function for that called `retrieve-fallback-descriptor`.  Should we change it there instead?  Are we intentionally ignoring `fallback-period-secs` for invalid descriptors?

File: waiter/integration/waiter/token_request_test.clj [hunk 2]
```diff
@@ -1397,3 +1397,49 @@
...
+      (-> token-description-1
+          (as-> token-description (post-token waiter-url token-description))
+          (assert-response-status 200))
+      (let [service-id-1 (retrieve-service-id waiter-url request-headers)]
+        (with-service-cleanup
+          service-id-1
+          (let [response-1 (make-request-with-debug-info request-headers #(make-request waiter-url "/hello" :headers %))]
+            (assert-response-status response-1 200)
+            (is (= service-id-1 (:service-id response-1))))
+          (-> token-description-2
```
Comment (sradack) [inline waiter/integration/waiter/token_request_test.clj]: Threading a small number of calls while also utilizing `as->` feels a bit awkward.  Would it be simpler to use `let` or write it normally?  e.g.  `(assert-response-status (post-token waiter-url token-description-2) 200))`

File: waiter/src/waiter/service_description.clj [hunk 3]
```diff
@@ -854,7 +854,7 @@
      If after the merge a permitted-user is not available, then `username` becomes the permitted-user."
     [{:keys [defaults headers service-description-template source-tokens token-authentication-disabled token-preauthorized]}
      waiter-headers passthrough-headers kv-store service-id-prefix username metric-group-mappings
-     service-description-builder assoc-run-as-user-approved?]
+     service-description-builder assoc-run-as-user-approved? throw-if-invalid?]
```
Comment (sradack) [inline waiter/src/waiter/service_description.clj]: Is a better approach to return `nil` when a service description would be invalid, and change the existing callers to throw on `nil`?

File: waiter/integration/waiter/token_request_test.clj [hunk 4]
```diff
@@ -374,10 +374,10 @@
                 services (json/read-str body)
                 service-tokens (mapcat (fn [entry]
                                          (some->> entry
-                                                  walk/keywordize-keys
```
Comment (sradack) [inline waiter/integration/waiter/token_request_test.clj]: Nitpick: Unrelated formatting changes throughout this file

File: waiter/src/waiter/service_description.clj (hunk 3 above)
Comment (shamsimam) [reply waiter/src/waiter/service_description.clj]: We still need metadata on why the service description is invalid to report in our error messages.

File: waiter/src/waiter/descriptor.clj (hunk 1 above)
Comment (shamsimam) [reply waiter/src/waiter/descriptor.clj]: > Are we intentionally ignoring fallback-period-secs for invalid descriptors?

Only the `fallback-period-secs` of the latest version of the token is considered. We ignore values of `fallback-period-secs` from previous versions of the token.

File: waiter/integration/waiter/token_request_test.clj (hunk 2 above)
Comment (shamsimam) [reply waiter/integration/waiter/token_request_test.clj]: Done.

File: waiter/src/waiter/descriptor.clj (hunk 1 above)
Comment (shamsimam) [reply waiter/src/waiter/descriptor.clj]: Moved the logic to `retrieve-fallback-descriptor`

File: waiter/integration/waiter/token_request_test.clj (hunk 4 above)
Comment (shamsimam) [reply waiter/integration/waiter/token_request_test.clj]: Undone.

File: waiter/src/waiter/service_description.clj (hunk 3 above)
Comment (sradack) [reply waiter/src/waiter/service_description.clj]: It could still return a map with partial results instead of nil.  I guess my larger point is that when it becomes desirable to have a function not throw in some cases, it seems like we should make it not throw in all cases, and rework the callers that expected an exception, as opposed to passing a parameter into the function causing it to throw in certain circumstances.  That seems cleaner to me.

**Comment by shamsimam** [comment]
@sradack ready for review

File: waiter/src/waiter/service_description.clj [hunk 5]
```diff
@@ -894,34 +894,43 @@
...
+         :service-preauthorized service-preauthorized
+         :source-tokens source-tokens})))
+
+  (defn validate-service-description
+    "Returns nil if the provided descriptor contains a valid service description.
+     Else it returns an instance of Throwable that reflects the validation error."
+    [kv-store service-description-builder
+     {:keys [core-service-description passthrough-headers service-description service-id waiter-headers]}]
+    (sling/try+
+      (let [stored-service-description? (fetch-core kv-store service-id)]
```
Comment (sradack) [inline waiter/src/waiter/service_description.clj]: The `?` on the end to me implies it's a boolean, but it's not.

Comment (shamsimam) [reply waiter/src/waiter/service_description.clj]: Removed the `?`

**Comment by shamsimam** [comment]
@sradack ready for review
